From 914872589fe80a1e92fb6238cc54e816097b3e60 Mon Sep 17 00:00:00 2001
From: Zhou Peng-B04994 <eagle.zhou@nxp.com>
Date: Mon, 13 Nov 2017 12:07:57 +0800
Subject: [PATCH 2784/5242] MLK-16806 - [i.MX8QXP/Malone]: segment fault in
 gdb environment

commit  2eba5900352b69e3843d717e230569a1ba716f1f from
https://source.codeaurora.org/external/imx/linux-imx.git

function wait_event_interruptible_timeout may return some special value
in gdb debug environment, so we need to add additional check to these
return values, otherwise, interrupt will reported wrongly

Signed-off-by: Zhou Peng-B04994 <eagle.zhou@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mxc/vpu-malone/Malone_Firmware/PAL/pal.c |   14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/drivers/mxc/vpu-malone/Malone_Firmware/PAL/pal.c b/drivers/mxc/vpu-malone/Malone_Firmware/PAL/pal.c
index ae7f21e..d75db10 100755
--- a/drivers/mxc/vpu-malone/Malone_Firmware/PAL/pal.c
+++ b/drivers/mxc/vpu-malone/Malone_Firmware/PAL/pal.c
@@ -590,6 +590,7 @@ MEDIAIP_FW_STATUS pal_qu_receive ( PAL_QUEUE_ID QuId,
 		void         *pMessage )
 {
 	u_int32 retval;
+	long ret;
 
 	ENTER_FUNC();
 	dprintf(LVL_FUNC, "QuId %d\n", QuId);
@@ -597,16 +598,23 @@ MEDIAIP_FW_STATUS pal_qu_receive ( PAL_QUEUE_ID QuId,
 		set_current_state(TASK_INTERRUPTIBLE);
 		schedule();
 	}*/
-	if(!wait_event_interruptible_timeout(irq_wq[QuId],
+	ret = wait_event_interruptible_timeout(irq_wq[QuId],
 			kfifo_len(&irq_fifo[QuId])>=4*sizeof(uint_addr)
 			/* || kthread_should_stop() */,
 			//uTimeoutMs
 			msecs_to_jiffies(uTimeoutMs)
-			))
-	{
+			);
+
+	if (ret == 0) {
 		dprintf(LVL_FUNC, "timeout %d ms\n", uTimeoutMs);
 		return MEDIAIP_FW_STATUS_TIMEOUT;
+	}	else if (ret == -ERESTARTSYS) {
+		dprintf(LVL_FUNC, "ERROR interrupted by a signal!!\n");
+		return MEDIAIP_FW_STATUS_FAILURE;
+	}	else {
+		dprintf(LVL_FUNC, "%ld returned from wait event\n", ret);
 	}
+
 	//if(kthread_should_stop())
 	//	return MEDIAIP_FW_STATUS_STOPPED;
 
-- 
1.7.9.5

